integration/test/metric/metric_value_query.go

```diff
@@ -0,0 +1,98 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+//go:build linux && integration
+// +build linux,integration
+
+package metric
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/amazon-cloudwatch-agent/integration/test"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch/types"
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+var metricValueFetchers = []MetricValueFetcher{
+	&CPUMetricValueFetcher{base: &baseMetricValueFetcher{}},
+}
+
+func GetMetricFetcher(metricName string) (MetricValueFetcher, error) {
+	for _, fetcher := range metricValueFetchers {
+		if fetcher.isApplicable(metricName) {
+			return fetcher, nil
+		}
+	}
+	err := fmt.Errorf("No metric fetcher for metricName %v", metricName)
+	log.Printf("%s", err)
+	return nil, err
+}
+
+type MetricValueFetcher interface {
+	Fetch(namespace string, metricName string, stat Statistics) ([]float64, error)
+	fetch(namespace string, metricSpecificDimensions []types.Dimension, metricName string, stat Statistics) ([]float64, error)
+	isApplicable(metricName string) bool
+	getMetricSpecificDimensions() []types.Dimension
+}
+
+type baseMetricValueFetcher struct{}
+
+func (f *baseMetricValueFetcher) fetch(namespace string, metricSpecificDimensions []types.Dimension, metricName string, stat Statistics) ([]float64, error) {
```

Comment:
Some linters catch this, but this function doesn't actually reference `f` here, which makes me think that this function does not need to be directly tied to the MetricValueFetcher. What it looks like to me is that an instance of a MetricValueFetcher defines a set of dimensions that it allows for querying, and from there they all will make the same structured query with their specific dimensions. 

I think that this might be worthwhile to restructure, to pull out this `fetch` function. Maybe we don't need the `Fetch()` function on the fetcher interface, and just define `getMetricSpecificDimensions()`, and call this separate utility (sorry) function with that as a parameter value.

```go
cpuFetcher := CPUMetricValueFetcher{}

values, err := MyFetchUtilFunction(namespace, cpuFetcher.getDimensions(), metricName, stat)
```

Reply:
ew

Reply:
jk, so I was trying to figure out how to do in golang what I'd do in java for this logic. And yeah maybe in golang world we don't need this as a... struct member function ? (I understood this as similar to what java class member functions are). I still do like having a single struct that you can call Fetch on instead of getting cpu fetcher using a factory, then passing a function in there into another util method. That seems like too much extra implementation detail to be a good interface for future tests. Let's continue to discuss what each of us care about in this code and how to satisfy both?

Reply:
If the alternative in the below works, is it ok if we leave it this ?

Reply:
Adam's comment and suggestion satisfies my concern with this.